### PR TITLE
Fix fp16 by converting outputs back to FP32

### PR DESCRIPTION
**Pull request by sgugger**
As was pointed out in #101, there is currently a bug in mixed precision training: the outputs are properly computed in mixed precision but are returned in FP16, and the loss computation is not inside a `torch.cuda.amp.autocast` context manager so is executed in full FP16, which is generally unstable (especially for softmax).

This was not discovered with Transformers models as the loss is computed inside the model, which is generally a better idea if one wants to use mixed precision with Accelerate.

To fix the problem, this PR:
- automatically converts the outputs of the model to FP32 so that the loss is executed in full precision (generally a better idea)
- adds a context manager `accelerator.autocast` for more complex loss functions that should be executed in mixed precision (as with all things accelerate, this context manager always work, it just does nothing if FP16 is not activated).

### CODE REVIEW

File: src/accelerate/accelerator.py [hunk 1]
```diff
@@ -550,3 +552,17 @@ def get_state_dict(self, model):
                 state_dict[k] = state_dict[k].float()
 
         return state_dict
+
+    @contextmanager
+    def autocast(self):
+        """
+        Will apply automatic mixed precision inside the block inside thics context manager, if it is enabled. Nothing
```
Comment (LysandreJik) [inline src/accelerate/accelerator.py:559]:
```suggestion
        Will apply automatic mixed-precision inside the block inside this context manager, if it is enabled. Nothing
```

**Review by LysandreJik — APPROVED** [review]
Looks great! Cool workaround on `autocast`. LGTM!
